h264/bits/bitreader.go

```diff
@@ -0,0 +1,119 @@
+/*
+DESCRIPTION
+  bitreader.go provides a bit reader implementation that can read or peek from
+  an io.Reader data source.
+
+AUTHORS
+  Saxon Nelson-Milton <saxon@ausocean.org>, The Australian Ocean Laboratory (AusOcean)
+
+LICENSE
+
+  Copyright (c) 2009 The Go Authors. All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are
+  met:
+
+    * Redistributions of source code must retain the above copyright
+  notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above
+  copyright notice, this list of conditions and the following disclaimer
+  in the documentation and/or other materials provided with the
+  distribution.
+    * Neither the name of Google Inc. nor the names of its
+  contributors may be used to endorse or promote products derived from
+  this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+// Package bits provides a bit reader implementation that can read or peek from
+// an io.Reader data source.
+package bits
+
+import (
+	"bufio"
+	"io"
+)
+
+// BitReader is a bit reader that provides methods for reading bits from an
+// io.Reader source.
+type BitReader struct {
+	r    *bufio.Reader
+	n    uint64
+	bits uint
+}
+
+// NewBitReader returns a new BitReader.
+func NewBitReader(r io.Reader) *BitReader {
+	byter, ok := r.(*bufio.Reader)
+	if !ok {
+		byter = bufio.NewReader(r)
+	}
+	return &BitReader{r: byter}
```

Comment:
If you are doing it this way, it can be `return &BitReader{r: bufio.NewReader(r)}` since `bufio.NewReader` doesn't re-wrap if not necessary. If you want to avoid unnecessary wrapping of other types that satisfy the necessary interface, then make the condition above assert on `.(bytePeeker)` where `bytePeeker interface { io.ByteReader; Peek(int) ([]byte, error) }`.

Reply:
What is the benefit of avoiding unnecessary wrapping ?

Reply:
If I have a `*bufio.Reader`, what is the value of wrapping it in an `*bufio.Reader`? You are already doing this check above, but that check is done in `bufio.NewReader`. However, if you have another type that implements the two methods we need here, what's the point of wrapping here?

Reply:
What I'm asking is if I just do `return &BitReader{r: bufio.NewReader(r)}`, and NewReader re-wraps something it doesn't need to, is there much downside to that in comparison to the downside of having more code to define that interface(`bytePeeker interface { io.ByteReader; Peek(int) ([]byte, error) }`) and do that check?

Reply:
They are orthogonal issues.

1. [`bufio.NewReader(r) == bufio.NewReader(bufio.NewReader(r))`](https://play.golang.org/p/-GNIccRILwD) due to the [checks in `bufio.NewReader`](https://golang.org/src/bufio/bufio.go?s=1414:1464#L36) as mentioned above. So, it is not being rewrapped with or without the check.

2. If you have a type that does what you want, and has the two needed methods, and you don't want it to be wrapped in a `*bufio.Reader` for whatever reason the code here will not allow that. A case for this might be a use-case-specific buffer that has has `io.ByteReader` and `Peek...`

Now. Either you just accept that everything is going to be a *bufio.Reader` here and you don't have control of the buffering rate, in which case you have just a single line return and no conditions in this code, or you have the check and define the interface to assert on.

I prefer the latter. More code here is not an issue.